espsecure/esp_hsm_sign/__init__.py

```diff
@@ -37,6 +38,12 @@ def read_hsm_config(configfile):
         if not config.has_option(section, option):
             raise configparser.NoOptionError(option, section)
 
+    # Check if config.["credentials"] is set to "prompt" and prompt for HSM PIN
+    credentials_config_val = config.get(section, "credentials")
+    if credentials_config_val == "prompt":
+        hsm_pin = getpass("\nPlease enter the PIN of your HSM:\n")
```

Comment:
Is there a way to pass this off directly to the HSM library instead of using getpass? From a security perspective, this would mean we don't have to worry about password management at all. That said, if that's not possible this approach is still much better than storage in a file.

I would suggest that we try to upstream this feature as it's good for the entire ecosystem.

Reply:
(I blame python) `getpass()` AFAIK has nothin to do with password, it is just `input()`that does not echo back what the user type, so I figure lets prompt it in one place and then store it in the `in-RAM` config, then downstream function can just use it.

The other incision point is in `establish_session` where `user_pin=config["credentials"]` but if establish_session is called a lot the poor `release engineer` will need to type pin A lot 😅 .

The sample usage is like this (my bad for not including it earlier):
```
python3 -m espsecure sign_data  --version 2 --hsm --hsm-config hsm-config.conf --output huhu-signed.bin huhu.bin
espsecure.py v4.7-dev
Padding data contents by 4091 bytes so signature sector aligns at sector boundary

Please enter the PIN of your HSM: <user-type-pin-here-with-no-echo>

Trying to establish a session with the HSM.
Session creation successful with HSM slot 0.
Trying to extract public key from the HSM.
Got public key with label esp32-secure-boot.
Connection closed successfully
Trying to establish a session with the HSM.
Session creation successful with HSM slot 0.
Got private key metadata with label esp32-secure-boot.
Signing payload using the HSM.
Signature generation successful.
Connection closed successfully
Pre-calculated signatures found
1 signing key(s) found.
```

Reply:
(on upstreaming) - of course, I wanted to pass the `local ACK` first before going upstream, this puppy is in it own commit so it should be good to go.

Reply:
So if I understand correctly, you don't want to have the user type the PIN twice? Is there a way to keep the session alive so we still have the HSM prompt directly?

Reply:
I wanted to override the content of `config.["credentials"]` with a prompt for the user to type-in the HSM PIN. this way if the script is later modified to do other things (like being able to sign multiple binaries, or do `sign-and-verify-in-one-shot`) the incision point already supports that because it is an earlier common point in the script.

I am not following the `Is there a way to keep the session alive so we still have the HSM prompt directly?` comment.
Why do we have to do that when the current solution already solves it?
Is there a more `secure HSM prompting` method than `getpass()` that I am missing?

Reply:
I guess if at all possible I would like to keep the PIN out of out control as if we consider it to have similar properties to a password or key, we should then probably be thinking about secure memory practices and whether the pin can be extracted. I'm probably overthinking this at this point though.

Reply:
(for posterity)
The primary concern here is to minimize `exposure time` of the `sensitive info`.

I found this REF https://python-pkcs11.readthedocs.io/en/latest/api.html#pkcs11.Token.open

Me an David discussed this and I tried these permutations:
```
    session = token.open(rw=True, user_pin=PROTECTED_AUTH)
NameError: name 'PROTECTED_AUTH' is not defined

    session = token.open(rw=True, so_pin=PROTECTED_AUTH)
NameError: name 'PROTECTED_AUTH' is not defined

    session = token.open(rw=True, PROTECTED_AUTH=config["credentials"])
TypeError: open() got an unexpected keyword argument 'PROTECTED_AUTH'
```

To see if we can minimize the `time to live` of the HSM PIN down to the minimum it needs to exist in RAM, unfortunately no luck on this front.

I also queried google and chatGPT to see if there is a known best practice to `sanitize` data after usage, the short summary is this.
```
It's important to note that secure memory wiping is highly dependent on the underlying system 
and memory management techniques. 
In general-purpose programming languages like Python, it's challenging 
to achieve complete control over memory deallocation and wiping. 
If you have stringent security requirements, using lower-level languages 
or leveraging specialized hardware might be more appropriate.
```

so at least we are not missing any cheap and obvious way to make this better.